### ISSUE TITLE
When current instrument is locked, clicking on an instrument in the mixer should still change current instrument

### ISSUE THREAD

**Issue by kmatheussen** [kmatheussen/radium#1412]
1. Start Radium
2. Open instrument widget
3. Click the "lock" icon in the upper left corner
4. Click on a different instrument in the mixer
5. Current instrument is not changed.

Also, after fixing this, the "lock" icon functionality should probably be enabled by default. It's usually annoying that the program changes current instrument for you. Perhaps current instrument should only change if clicking on it in the mixer, or manually changing current track in the editor (i.e. not when it's changing because we play a song and a new current track has a different instrument when playing a new block). Then we don't need the "lock" icon.

**Comment by yust1n** [comment]
A fix for this will be commited soon. :)

**Comment by yust1n** [comment]
Done.

**Comment by kmatheussen** [comment]
And instrument should perhaps change when explicitly changing editor track. Or at least it should be possible to configure the program to do so.

**Comment by yust1n** [comment]
Your patch seems to work.

**Comment by kmatheussen** [comment]
This commit seems to ensure current instrument is changed when changing current editor track with the keyboard: 

https://github.com/kmatheussen/radium/commit/b520d2d2ff853bdacb99636eec7d9a70d7100bf7

Any other situations now? Hopefully it's possible to remove the "locked instrument" option soon. I don't think anyone wants the non-locked option anymore...

**Comment by kmatheussen** [comment]
I guess it makes sense for  the "lock" icon in the upper left corner to be replaced with a "switch current instrument" icon now...

**Comment by kmatheussen** [comment]
Problem introduced after last commit: Radium can (wrongly) change instrument automatically when moving the horizontal scrollbar in the editor.

**Comment by kmatheussen** [comment]
> Problem introduced after last commit: Radium can (wrongly) change instrument automatically when moving the horizontal scrollbar in the editor.

Also fixed.

**Comment by kmatheussen** [comment]
Replaced locked checkbox with select button. TODO: Fix right-clicking the select-button.

**Comment by kmatheussen** [comment]
TODO: Can't select midi instruments from the select-button.

**Comment by kmatheussen** [comment]
TODO: Immediately set current instrument to the instrument of the current track in the editor when moving the cursor in the editor, not just when switching between tracks.

**Comment by kmatheussen** [comment]
TODO: Clean up code, replace all the "even_if..." variables with "also_set_current_instrument" variables, etc.

**Comment by yust1n** [comment]
In file included from Qt/mQt_patch_widget_callbacks.h:10,
                 from Qt/Qt_instruments.cpp:180:
Qt/Qt_patch_widget_callbacks.h: In constructor 'Patch_widget::Patch_widget(QWidget*, Patch*)':
Qt/Qt_patch_widget_callbacks.h:206:5: error: 'locked_instrument' was not declared in this scope; did you mean 'make_instrument'?
  206 |     locked_instrument->_hovered_callback = [is_alive](bool do_enter){
      |     ^~~~~~~~~~~~~~~~~
      |     make_instrument
make: *** [Makefile:1661: /tmp/radium_objects_makepkg/Qt_instruments.o] Error 1
          Radium  compiled. (6s)
==> ERROR: A failure occurred in build().
    Aborting...
Another TODO :D

**Comment by kmatheussen** [comment]
Fixed

**Comment by yust1n** [comment]
Outch. Please remove my last commit.
I was still testing before committing and didn't expect you to be around.

**Comment by kmatheussen** [comment]
No it's fine. :-) That code is not going to be used anyway. I just left it
there (#if 0-ed out) so that it could serve as a template for what to do
when hovering the "select instrument" button, which hasn't been implemented
yet.

**Comment by yust1n** [comment]
MIDIinstruments also don't get displayed in the modularmixer :(

**Comment by kmatheussen** [comment]
TODO: Remove all "force as current instrument" menu options.

**Comment by kmatheussen** [comment]
> MIDIinstruments also don't get displayed in the modularmixer :(

Yeah, a MIDI instrument is a different type of thing. It doesn't make sense adding MIDI instruments in its current form into the modular mixer. What should be done instead is to write new "Audio instrument"s that handle MIDI. Afterwards (in the code) remove "MIDI instrument", and rename "Audio instrument" to just "Instrument".

**Comment by kmatheussen** [comment]
TODO: Update manual.

**Comment by kmatheussen** [comment]
Update: all "TODO"s are done, except this one:

"TODO: Clean up code, replace all the "even_if..." variables with "also_set_current_instrument" variables, etc.".

But that's just an internal thing. It should be ready for release now...

**Comment by kmatheussen** [comment]
TODO: Check that the program does the right things (whatever that might be) when deleting and creating instruments.

**Comment by yust1n** [comment]
Hmmm ... maybe selecting a track in the editor via mouse should also change the instrument ?
Guess I need to learn some scheme.

**Comment by kmatheussen** [comment]
Yes, adding it now.
TODO: Selecting a track in the editor via mouse should probably change current instrument.

**Comment by kmatheussen** [comment]
> Guess I need to learn some scheme.

The big advantage of writing code in scheme in Radium is that you don't have to compile anything when doing changes. You don't even have to restart Radium.

For instance, when fixing a bug in a function, first I edit the text in emacs, and then - while the cursor is still placed inside the function I edited - I press Ctrl+Alt+x. That's it.

To make emacs behave like, it's only necessary to add this line to ~/.emacs:
`(setq scheme-program-name  "/path/to/radium/bin/s7webserver/s7webserver_repl.py")`

**Comment by yust1n** [comment]
I use vim .)
So in other words it just saves the edited scm.file and executes webserver_repl.py ?

**Comment by kmatheussen** [comment]
> I use vim .) So in other words it just saves the edited scm.file and executes webserver_repl.py ?

No, webserver_repl.py is a REPL. It's a program, running all the time, that reads from stdin and replies to stdout. Ctrl+Alt+x only sends the block of code the cursor is currently placed in to the REPL. If you want to evaluate the whole file instead, the shortcut to use is ctrl+c and afterwards press l. I believe this is the most popular way to do programming in Lisp. I'm sure you can do this in VIM too.

**Comment by kmatheussen** [comment]
At one point in my twentieth I tried to switch to VIM, but I gave up after about a month. I had been too accustomed to emacs so it wasn't possible to make the switch. :-)

**Comment by kmatheussen** [comment]
Update: All TODOs, except the internal one, are finished. May be ready for release again.

**Comment by yust1n** [comment]
Found an issue when asigning an existing instrument to a track in the editor.
Cursor needs to be moved to show the selected instrument.

**Comment by yust1n** [comment]
https://iainctduncan.github.io/learn-scheme-for-max/part_1.html
This seems a nice place to start with scheme/lisp.

**Comment by kmatheussen** [comment]
Thanks, adding:

TODO:
1. Switch current instrument when assigning an instrument to a track in the editor.
2. Editor track header doesn't update when changing current instrument for track. (Recipe: 1: Assign a new instrument to a track. 2. Assign an existing instrument to the same track. 3. Instrument is assigned, but the header is not uptodate.)
3. Current instrument should probably change when clicking trackname in the editor. I purposedly didn't implement this, but I see now that it should probably change current instrument

> https://iainctduncan.github.io/learn-scheme-for-max/part_1.html
Yes, very well-written.

**Comment by kmatheussen** [comment]
TODO: Add recipe to the manual on how to develop in scheme in Radium using emacs. (What to add to ~/.emacs, how to start the repl (I.e. M-x run-scheme), how to evaluate code)

**Comment by yust1n** [comment]
I start disliking this new feature. It's pretty annoying while editing when the instrument stays on an instrument of the block before that doesn't even exist in the actual block.

**Comment by kmatheussen** [comment]
> It's pretty annoying while editing when the instrument stays on an instrument of the block before that doesn't even exist in the actual block.

Sorry, I don't understand this. Are you talking about mouse or keyboard?

> Maybe it's better to much this stuff to an experimental branch until it works proper and doesn't interrupt the workflow like it does at the moment.

Ouch, sure we can't fix the annoying parts?

**Comment by kmatheussen** [comment]
I actually thought the old behavior was annoying, so I wanted to fix that. But if this is more annoying, then obviously it shouldn't be included in the next release.

**Comment by kmatheussen** [comment]
To clarify, what I found very annoying before was that the current instrument was changed when playing a song, so that the current instrument always was the same as the current track in the currently playing block in the current seqtrack.

Everything else should perhaps be as before, as you suggest now, but I also want to get rid of the "lock instrument" button in the instrument widget, and replace it with a "select instrument" button, as has been done.

I hope we don't have to revert the code. I don't think it is that much work to just manually fix the annoying parts.

**Comment by yust1n** [comment]
So only lock the instrument while playing a song ?
Maybe also keep the lockinstrumentbutton and make it an instrumentchooser by rightclicking it ?
This way the transition would hurt less ;D

**Comment by kmatheussen** [comment]
Excellent. Thanks for raising this before release.

**Comment by yust1n** [comment]
I found other stuff that should also be fixed like the trackmuter in the sequencer. It changes when you are hovering over another block. Unfortunatelly I had more than one sequencertrack and I had to move that window close to the block I wanted to set the mutes.
https://github.com/kmatheussen/radium/issues/1049#issuecomment-1986880404
These options are hidden pretty well. I would love to be able to have an easier access.
"Play notes under the cursor when moving up or down" is also hard to find/access.
I bet this is all hidden somewhere in scheme.
So take this as a reminder for me where to start searching :)

**Comment by kmatheussen** [comment]
> .sf2 Works in Drums but I want to see waveforms :)

A quick fix for that is to use the sample player instrument to load the drums.sf2 file (yes, the sample player instrument supports sf2!). Fluidsynth does some magic that often makes sf2 files sound much better though (probably a different type of resampler algorithm), but maybe it doesn't sound that much better for drums.

**Comment by kmatheussen** [comment]
Hmm, seems to be bugs in the sample mapping of the sample player instrument when using drums.sf2.

**Comment by yust1n** [comment]
Whenever I tried it there was only one sample being played.
Same with xiinstruments.
OMG. I rolled back to https://github.com/kmatheussen/radium/commit/822cb9ab3d1b5205351e54bdcd06c07979a74ccd
and just discovered that you already had the instrumentchooser hidden in a submenu on rightclick.

**Comment by yust1n** [comment]
There is also a problem with stereosamples in a soundfont. Pitch is too low. Atleast the slicedbeats I created with shuriken import now fine in Radium when I convert them into mono and use only one instrument.

**Comment by kmatheussen** [comment]
I've fixed the code now so that it always changes current instrument when switching block/track, except when playing song. I hope it's not necessary to put back the "lock instrument" functionality. It's quite efficient to switch current instrument now by right-clicking the "switch instrument" button and assigning a shortcut, so that might be an okay workaround. But if not, it's still possible to revive "lock instrument".

**Comment by yust1n** [comment]
Did a quickcheck. The sampleplayer doesn't show the loaded sample at all ( not even in another track like before ).

**Comment by kmatheussen** [comment]
Can you provide a recipe? Because I don't know what you are talking about. When doesn't the sample player not show the loaded sample?

**Comment by yust1n** [comment]
Just add a new sampleplayer and load a sample.

**Comment by yust1n** [comment]
When adding another sampleplayer the sample loaded changes the trackname of the sampleplayer to the left.

**Comment by kmatheussen** [comment]
> Just add a new sampleplayer and load a sample.

Seems to work fine. What part of the program are you thinking of when you say "doesn't show the loaded sample"?

**Comment by kmatheussen** [comment]
> When adding another sampleplayer the sample loaded changes the trackname of the sampleplayer to the left.

Can you provide a recipe? I don't understand what you mean.

**Comment by yust1n** [comment]
Did a short screenrecording.

**Comment by kmatheussen** [comment]
Thank you, should be fixed.

**Comment by yust1n** [comment]
> Could not find "711620__joshstovall__cymbal-roll.wav" in ".".

That's what I get when I try to load that new demosong.

**Comment by kmatheussen** [comment]
fixed

**Comment by yust1n** [comment]
Nice. Even when on a track and changing to editmode the instrument in the actual editortrack gets used and when disabling the editmode the original instrument gets played. Didn't recheck the demosong but the rest looks fine at the moment.

**Comment by kmatheussen** [comment]
Great. If there's still situations where current instrument is not changed
when it should, you can probably fix it brute-forcely by applying this
patch, instead of reverting all changes:

diff --git a/Qt/Qt_instruments.cpp b/Qt/Qt_instruments.cpp
index 1742c6c0c..bf8fc2fc4 100644
--- a/Qt/Qt_instruments.cpp
+++ b/Qt/Qt_instruments.cpp
@@ -849,7 +849,7 @@ static void GFX_PP_Update_internal(struct Patch *patch,
bool is_loading, bool op

   if (open_even_if_locked == false){
     printf("Gakk\n");
-    //open_even_if_locked = true;
+    open_even_if_locked = true;
   }

 #if 0

And to fix it properly, it's probably just putting a breakpoint inside that
if-test and see where the wrong argument came from.

**Comment by yust1n** [comment]
Wow ... found another bug related to this topic.
When you select a seqtrack in the modularmixer there is also an instrumenteditor for it.
It still changes :(
Maybe seqtracks should also be added to the instrumentlist.

**Comment by kmatheussen** [comment]
> When you select a seqtrack in the modularmixer there is also an instrumenteditor for it. It still changes :

I'm not sure if I understand the problem. Is this recipe correct?

1. Start Radium
2. Create a new audio seqtrack
3. Show the instrument widget
4. Click the new audio seqtrack object
5. The new audio seqtrack object is shown in the instrument widget.

If so, isn't that the correct behavior?

**Comment by kmatheussen** [comment]
> Seems there is no way to turn off granulation in seqtracker .... hmmm.

There is no other way to stretch a sound in Radium. It uses granulation. If you turned off granulation, there would be no way to stretch an audio file. If an audio file is not stretched, granulation is automatically turned off. I don't understand the problem...

**Comment by kmatheussen** [comment]
Maybe you're using the "Pitch" function (when you double-click a sound file), but what you are really looking for is the "Speed" function (the "red" square in the soundfile)?
<img width="640" alt="Screenshot 2024-03-26 at 17 24 58" src="https://github.com/kmatheussen/radium/assets/1146279/95083b5b-3eef-4646-9ec8-ac9755674213">

**Comment by yust1n** [comment]
The problem with switching instruments is solved. In the end it happend only when looping in the sequencer so adding !isSeqlooping made it behave nicely.
The AudioSeqtrack kindof should behave like the sampleplayer. I will find a way to add a button to turn granulation off manually and all is fine :) Will be some pratice on how to deal with scheme for me.
Thanks for repkying.

**Comment by kmatheussen** [comment]
I mean, to turn off granulation you just right-click one of the
stretch-handlers in the audio file and select "Reset". That turns off both
stretching and granulation. Isn't that what you want?

**Comment by yust1n** [comment]
Indeed, that's it \o/
So the button should be "No stretching when pitching" :D
THANK YOU !!!

**Comment by kmatheussen** [comment]
In Radium, "pitch" means only changing pitch, not the length of the sample.
I think that's pretty common. Only changing sample rate, which you seem to
want, is called "speed" in radium.

**Comment by kmatheussen** [comment]
I guess there is an interface issue here. The "speed" handlers only show
multiplication of default sample rate of the audio file. Maybe it also/or
should show what this multiplication means in terms of pitch change.
E.g. instead of displaying "2.0", it could alternatively display "+12"
(i.e. one octave up).

**Comment by yust1n** [comment]
I just followed the logic of a sampleplayer where pitching up a sample also results in a shorter samplelength ( like a recordplayer ... hmm wait ... that's also speed but I remember it being called pitch ).
Now that I understood the logic of the audioseqtracker everything is fine :) It didn't came to my mind that it's different.
Radium has so many hidden features that it just takes a while to get them all. Like the automation in the MainBus. That's what I was looking for to do automation except that it lacks the possibility to use midicontrollers ( would be perfect if it did ).

**Comment by kmatheussen** [comment]
Yeah that's true. But the sample player is mostly used to play notes,
while in the sequencer, audio samples are mostly used to play
sequences of notes/etc. (at least that's my assumption)
For the sample player, we can easily cheat by only changing sample
rate to change pitch, because then you get the best sound and the
lowest CPU usage. But for the sequencer,
if you are singing in the wrong pitch, and you want to fix that by
changing the pitch in post-production, I would think you normally
expect the duration of the sample to stay the same.

**Comment by kmatheussen** [comment]
> Radium has so many hidden features that it just takes a while to get them all. Like the automation in the MainBus. That's what I was looking for to do automation except that it lacks the possibility to use midicontrollers ( would be perfect if it did ).

You mean if you right-click inside the seqblock area of the main bus,
and select "New automation..."?
If so, when you say that it lacks the possibility to use
midicontrollers, do you mean record automation? (I agree that would
probably be nice)

**Comment by yust1n** [comment]
> If so, when you say that it lacks the possibility to use
> midicontrollers, do you mean record automation? (I agree that would
> probably be nice)
That's what I did with PD in the end but actually don't use it.
Doing some oldschool dnb at the moment just to learn how to use everything the trackerway.

**Comment by yust1n** [comment]
![automation_paste_bug](https://github.com/kmatheussen/radium/assets/44263445/ce43e8ff-be70-4691-9a20-43253d112aa5)
When the start off the automation gets extended like the yellow line it's possible to paste.

**Comment by kmatheussen** [comment]
On Wed, Mar 27, 2024 at 4:08 PM yust1n ***@***.***> wrote:

> If so, when you say that it lacks the possibility to use
> midicontrollers, do you mean record automation? (I agree that would
> probably be nice)
> That's what I did with PD in the end but actually don't use it.
> Doing some oldschool dnb at the moment just to learn how to use everything
> the trackerway.
>
>
You can record automation from a MIDI controller in editor blocks though,
so I guess one of the best workarounds would be to have dedicated editor
seqtracks for automation.
And if you need really long automation, just take a normal editor seqlock
and stretch it (using the stretch handlers) so that the seqblock becomes
equally long as your song (much quicker
than changing the number of lines in the editor block/etc.).

**Comment by kmatheussen** [comment]
On Fri, Mar 29, 2024 at 9:58 AM yust1n ***@***.***> wrote:

> automation_paste_bug.png (view on web)
> <https://github.com/kmatheussen/radium/assets/44263445/ce43e8ff-be70-4691-9a20-43253d112aa5>
> When the start off the automation gets extended like the yellow line it's
> possible to paste.
>
>
Can you make a recipe on how to provoke this bug?

**Comment by yust1n** [comment]
Create a new automation, add a new node and pull it to the left so it's at the same timemarker as the beginning of the automation. Copy automation and paste it elsewhere.

**Comment by yust1n** [comment]
> You can record automation from a MIDI controller in editor blocks though
Did you try recording more than one MIDI controller at once like filtercutoff and filterresonance ?

**Comment by kmatheussen** [comment]
On Fri, Mar 29, 2024 at 11:54 AM yust1n ***@***.***> wrote:
>
> Create a new automation, add a new node and pull it to the left so it's at the same timemarker as the beginning of the automation. Copy automation and paste it elsewhere.
>

Thanks!

**Comment by kmatheussen** [comment]
On Fri, Mar 29, 2024 at 11:58 AM yust1n ***@***.***> wrote:

> You can record automation from a MIDI controller in editor blocks though
> Did you try recording more than one MIDI controller at once like
> filtercutoff and filterresonance ?
>
>
There shouldn't be any limitations on how many automations you can record
at once.

**Comment by yust1n** [comment]
```
 On Fri, Mar 29, 2024 at 11:58 AM yust1n ***@***.***> wrote:
>
> You can record automation from a MIDI controller in editor blocks though
> Did you try recording more than one MIDI controller at once like filtercutoff and filterresonance ?


 There shouldn't be any limitations on how many automations you can record at once.
https://users.notam02.no/~kjetism/radium/documentation/index.html?page=effect_popup_menu#record

```
Ahhh, I need to activate every controller I want to record every time _before_ Radium can record it.
That's pretty complicate. Additionally EditMode must be turned on.
Nice thing is that it really let's you record automation for any other instrument in the block.

**Comment by yust1n** [comment]
> `Would be good if you could reproduce the crash.``

I am not in the mood to crash my whole system atm so you got to be patient ;D

**Comment by yust1n** [comment]
> Instruments: Fix current instrument changing
> automatically when playing song and switching
> between blocks. This was only a bug in the official
> binary releases of Radium 7.3.84.

Sorry didn't commit yet when you when you announced a new release which I discovered today.
It wasn't a big thing and should also happen on osx since it only occured when playing in loopmode in the sequencer.
At the moment I am studying your code trying to implement a "No stretch"-button just for the fun of knowing how to do it. <ra> and <gui> in scheme are still a mysterium.
Is there any reason why I don't see a "7.4.76"-tag on github ?

**Comment by kmatheussen** [comment]
On Mon, Apr 1, 2024 at 4:08 PM yust1n ***@***.***> wrote:

> Instruments: Fix current instrument changing
> automatically when playing song and switching
> between blocks. This was only a bug in the official
> binary releases of Radium 7.3.84.
>
> Sorry didn't commit yet when you when you announced a new release which I
> discovered today.
> It wasn't a big thing and should also happen on osx since it only occured
> when playing in loopmode in the sequencer.
>

The bug was caused by the source code on the build machine being different
from the repository on github. The reason it didn't manifest on mac/arm is
because that version is built on a different computer.



> At the moment I am studying your code trying to implement a "No
> stretch"-button just for the fun of knowing how to do it. and in scheme are
> still a mysterium.
> Is there any reason why I don't see a "7.4.76"-tag on github ?
>
>
Forgot it. Fixed now.

**Comment by yust1n** [comment]
PD and Faust disappeared from the instrumentlist.

**Comment by kmatheussen** [comment]
On Mon, Apr 1, 2024 at 5:36 PM yust1n ***@***.***> wrote:
>
> PD and Faust disappeared from the instrumentlist.
>

I can't find anything wrong. Which instrument list are you talking about?

**Comment by yust1n** [comment]
The one in the editor ( click me ) and the one in the modularmixer.

**Comment by kmatheussen** [comment]
Oh, you mean the "New Pd instrument" entry, etc. Well, the menu was
reorganized on purpose a while ago. I removed some entries I thought were
too esoteric to be placed directly in that menu.

**Comment by kmatheussen** [comment]
You can add keyboard entries for Instruments you use a lot. For instance to
add keybinding to create Pd instrument:
1. Click the "(click me)" button
2. Select "from instrument/effect-menu..."
3. Right-click Pd -> Simple Sine Instrument and select "New keybinding".

**Comment by kmatheussen** [comment]
Don't know what happened, but  a "git push" seems to have fixed it.

**Comment by kmatheussen** [comment]
On Mon, Apr 1, 2024 at 11:01 PM Kjetil Matheussen ***@***.***>
wrote:

> Oh, you mean the "New Pd instrument" entry, etc. Well, the menu was
> reorganized on purpose a while ago. I removed some entries I thought were
> too esoteric to be placed directly in that menu.
>
>

But you do raise a concern that the Pd And Faust features are becoming too
hidden deep inside the program after that change.
One strange way to make it more visible would be to add an "Esoteric
features" menu to the manual,  but perhaps there are other/better ways.

**Comment by yust1n** [comment]
These are built-in synths/tools and it feels pretty strange when they are not listed in the popupmenu.
